code/modules/keybindings/setup.dm

```diff
@@ -0,0 +1,34 @@
+// Set a client's focus to an object and override these procs on that object to let it handle keypresses
+
+/datum/proc/key_down(key, client/user) // Called when a key is pressed down initially
+	return
+/datum/proc/key_up(key, client/user) // Called when a key is released
+	return
+/datum/proc/keyLoop(client/user) // Called once every server tick
+	set waitfor = FALSE
+	return
+
+/// Set mob's focus
+/// TODO - Do we even need this concept?
+/mob/proc/set_focus(datum/new_focus)
+	if(focus == new_focus)
+		return
+	focus = new_focus
+
+/// Turns a keys bitfield into text showing all bits set
+/proc/keys2text(keys)
+	if(!keys)
+		return ""
+	var/list/out = list()
```

Comment:
See also above. Don't actually know if list operations are cheaper than string operations but for a completely self-contained set of strings, generating a list and then concatenating that list feels a bit excessive.

Reply:
Byond strings are cstrings, which are immutable. Concatenating many together in sequence means having the initial string, then for each chunk added creating the string to add, creating the concatenated result, and then discarding the original and the add string.

List and Join can be dramatically cheaper in this style of use, which is effectively a budget StringBuilder. Most of the intermediate allocation and freeing is skipped, and the final join is done as a single act from the combined set length, the length of the members, and the joining particle.